### PR TITLE
Start of centralizing node writes via nodeWriter

### PR DESCRIPTION
**Pull request by cgwalters**
Part of https://github.com/openshift/machine-config-operator/pull/3135

---

daemon: Have nodeWriter maintain ref to lister and node name

The NodeWriter is an instance of the "actor" pattern effectively;
it processes messages to update its internal state.

However, it's a bit redundant to have every update take the node
lister and node name - we only use this on the daemon side,
where there is exactly one node we will be updating.

Have the clusterNodeWriter instance of this interface cache
references to that data and avoid passing it on every message.

Prep for reworking the NodeWriter to have its own clientset.

---

controller: Export default resync period function

So I can use this from the daemon code.

---

daemon: Have nodewriter use kubelet credentials

---

daemon: Change hypershift path to use nodewriter

Ensure that this code path is using kubelet credentials.

---

### CODE REVIEW

**Comment by cgwalters** [comment]
Digging into this, there's a fair amount of stuff that's needed:

- Move all Event notifications to use nodeWriter
- Fix up a few places that are calling `setAnnotations` directly to use nodeWriter
- Drop privs from the RBAC

However, I think after that it would really make the most sense to change nodeWriter into "nodeActor" and have it also be the sole thing providing *read* access to the node object to the rest of the MCD.  This would obviously touch more code.  But one immediate thing this would clean up is that today the MCD (i think) watches *all* nodes via informers, which is clearly just wrong.  We should only watch our own node, and then the nodeActor provides a "node changed" event to the rest of the MCD.

**Comment by cgwalters** [comment]
Ah yes, I forgot to start the informer; classic.  Will look at fixing that up.

**Comment by cgwalters** [comment]
Does this direction make sense to everyone?

**Comment by cgwalters** [comment]
> One option is to just add that power - then we use kubelet creds for everything.  It's a bit tempting.

I looked at this briefly; I think it'd require us to carry a patch to the apiserver to do that.  Definitely ugly.  I think I'd lean towards going this direction instead, but open to debate.

**Comment by cgwalters** [comment]
/approve cancel
per my SOP for big changes - big PRs should get another approver and a lgtm.

File: pkg/daemon/writer.go
Comment (kikisdeliveryservice) [inline pkg/daemon/writer.go:104]:
```suggestion
	glog.Infof("Starting Node Informer")
```

**Review by kikisdeliveryservice — COMMENTED** [review]
As a first pass (without reviewing the discussion point on the PR), I think this is a solid optimization that yields more readable and centralized code. Always felt odd to have to pass nodenames through daemon functions bc well.. shouldn't it know where it is? :laughing:

**Comment by cgwalters** [comment]
OK cool.  I took the first two commits and put them in https://github.com/openshift/machine-config-operator/pull/3143 - I think we can get that in now. 

I'll add some more work on top of this one for the followup [listed above](https://github.com/openshift/machine-config-operator/pull/3141#issuecomment-1121728037).

File: pkg/daemon/writer.go
Comment (yuqi-zhang) [inline pkg/daemon/writer.go:31]: I like the idea of using kubelet's kubeconfig just for node writes as we do here. The MCD can still have node list/get/watch perms used everywhere else

Comment (yuqi-zhang) [reply pkg/daemon/writer.go:31]: Sure, mostly had a question on https://github.com/openshift/machine-config-operator/pull/3143/files#r870410935

I'd also like to try to rebase my PR on this, so was wondering which of the two directions you were thinking of taking

File: pkg/daemon/writer.go
Comment (cgwalters) [reply pkg/daemon/writer.go:68]: Currently this is just dead code - I think we probably should change the daemon to use this to *read* the node too but it's a more invasive change than I want to try to land as one unit, and not on the critical path (see https://github.com/openshift/machine-config-operator/pull/3143#discussion_r870503417 )

What took me a while to debug is that the informer won't actually return "I started" unless you attach an event handler.

I'm only creating an informer here because `internal.UpdateNodeRetry` wants one, but I could change that to support doing a direct client `Get()`.

OTOH...hmmm...we may suffer from races unless the daemon and nodeWriter actually share a common lister state for the node object.

Going to plow forward with just "phase 0" but if we find races we may need to go to "phase 1".

File: cmd/machine-config-daemon/start.go
Comment (cgwalters) [reply cmd/machine-config-daemon/start.go:245]: (Forgot to look at this one, will do in the next pass)

**Comment by cgwalters** [comment]
OK, let's see how this goes in a CI run; only compile tested locally.

**Review by yuqi-zhang — APPROVED** [review]
Approving the general direction. Also rebased https://github.com/openshift/machine-config-operator/pull/3135 onto this

**Comment by kikisdeliveryservice** [comment]
```level=fatal msg=failed to fetch Cluster: failed to fetch dependency of "Cluster": failed to generate asset "Platform Quota Check": error(MissingQuota): compute.googleapis.com/n2_cpus is not available in us-central1 because the required number of resources (12) is more than remaining quota of 8
```
Trying this again, will report if see same failure
/test e2e-gcp-op

**Comment by kikisdeliveryservice** [comment]
new error, will report:

```level=error msg=Error: Error waiting for instance to create: Quota 'N2_CPUS' exceeded.  Limit: 24.0 in region us-central1.
level=error
```

**Comment by cgwalters** [comment]
> {Operator progressing (ProfileProgressing): Waiting for 1/6 Profiles to be applied  Operator progressing (ProfileProgressing): Waiting for 1/6 Profiles to be applied}

Hmm, this was a bit concerning at first because I thought NTO might be generating some machineconfig, but AFAICS that isn't the case.  I looked at their logs but didn't see the problem.
/test e2e-agnostic-upgrade

**Comment by yuqi-zhang** [comment]
The installer PR merged. Let's try

/test e2e-gcp-op

**Comment by cgwalters** [comment]
> E0513 12:40:14.615182    1711 writer.go:199] Marking Degraded due to: failed to cordon node (5 tries): [timed out waiting for the condition, cordon error: nodes "ci-op-bsr23qd1-1354f-4tpmg-worker-b-6dh2l" is forbidden: User "system:serviceaccount:openshift-machine-config-operator:machine-config-daemon" cannot patch resource "nodes" in API group "" at the cluster scope]

OK, I think we need to take this PR *together* with https://github.com/openshift/machine-config-operator/pull/3135

**Pull request closed without merging**
